core/typeSugar.ml

```diff
@@ -3000,20 +3113,31 @@ let rec type_check : context -> phrase -> phrase * Types.datatype * usagemap =
                   | ft ->
                       let rettyp = Types.fresh_type_variable (lin_any, res_any) in
                       begin
-                        unify_or ~handle:Gripers.fun_apply ~pos
-                                ((exp_pos f, ft), no_pos (`Function (Types.make_tuple_type (List.map typ ps),
-                                                                     context.effect_row, rettyp)))
-                                ((exp_pos f, ft), no_pos (`Lolli (Types.make_tuple_type (List.map typ ps),
-                                                                  context.effect_row, rettyp)))
+                        let funt = no_pos (`Function (Types.make_tuple_type (List.map typ ps),
```

Comment:
They're not recursive, I just decided to use `and`, as it allowed me to shave off a couple of `in`s. Sorry, just a bad habit on my part - happy to change if preferred?

Reply:
I'd never seen `let and` used like that; I guess it just struck me since I'd not seen it in the Links codebase before. Minor style comment, I'm happy with it staying

Reply:
I'd actually suggest using `in` as a matter of style here to document the non recursiveness. If you use `and` unnecessarily, then ocamlc (and humans of the future) may need to do unnecessary work to determine the functions actually aren't mutually recursive.

Reply:
There's no `rec` keyword, so ocamlc won't need to worry :). But that's a fair point about mental overhead - I'll change that now.